### DISCUSSION TITLE
Dynamic Axes with OpenVINO EP

### DISCUSSION THREAD

**Discussion by agrawalsourav98** [immich-app/immich#7406] [General]
I made a PR recently that fixed a bug related to loading external weights when using OpenVINO (see #7389). But recent testing has revealed that the fix works only in certain scenarios. I tested it using the following command to create a session,

```
session = onnxruntime.InferenceSession('model.onnx', providers=['OpenVINOExecutionProvider'],provider_options=[{'device_type': 'GPU_FP32'}])
```

The fix works when `device_type` is specified and `GPU_X` is specified. But immich currenlty uses this,
```
session = onnxruntime.InferenceSession('model.onnx', providers=['OpenVINOExecutionProvider'],provider_options=[{'device_id': 'GPU'}])
```
which doesn't work. 

The issue here seems to be that `batch_size` is only recognized in the first case and not in any other case.

There are 3 fixes for this that should work with every scenario,

1. Replace `batch_size` with `-1`. This would mean reverting to something similar that was done earlier (changing inputs and outputs) but would support batches greater than 1.
2. If we are sure, that inference would always be single batch (I am not sure how concurrency works currently), we can get rid of the dynamic axes stuff when exporting models to `hugging_face`.
3. Instead of using `device_id`, we use `device_type` in `provider_options` and anywhere there is no GPU device available, we just use the `CPUExecutionProvider` as there would be hardly any performance gain when using `OpenVINOExecutionProvider` with `CPU`

IMO the simplest would be 2 if we only do single batch inference or 1 which would retain similar behaviour as other EPs.

Previous discussion at #6869 

What do you think @mertalev?

**Comment by mertalev** [comment]
Thanks for looking more into this!

1. I don't see how this would fix it. Since `batch_size` is just a name for the dynamic axis, -1 shouldn't behave differently in this case.

2. I intend to add dynamic batching, so making all the models static isn't ideal.

3. This seems like the best solution to me. It's a small change that doesn't affect the exported models.

**Reply by agrawalsourav98** [reply]
I just ran inference and this is what I had observed. I will verify using openvino library that the dynamic batch is indeed passed as it and works as expected and then try to see if there is anything we can do about this. As, batching is something that we are interested in the future, it makes sense to verify once.

Yeah, 3 seems like an easy solution which works.

**Reply by mertalev** [reply]
That's interesting if `batch_size` doesn't work, but -1 does. That's definitely a bug.

Also yeah, it'd be great to test that the dynamic axis actually works 😅

**Reply by agrawalsourav98** [reply]
Yeah. I did check the openvino support matrix [here](https://docs.openvino.ai/archive/2023.1/openvino_docs_OV_UG_Working_with_devices.html#devicesupport-feature-support-matrix) and I think it would be a waste of time to pursue/debug this (as the support is partial and maybe that's why it does not with with device_id) . Implementing option 3 would mean it works with dynamic batching as well. It should work for integrated gpus just fine but dedicated gpus might be an issue which I cannot debug ATM as I don't have one.

**Reply by KoffeinKaio** [reply]
If you want to test stuff on a Arc A380 I can do that for you.
preferably in some docker way

**Reply by agrawalsourav98** [reply]
Yeah, that would be good. I would provide a simple python script if you run that inside the openvino docker image and get the output that would be helpful.

**Comment by agrawalsourav98** [comment]
@KoffeinKaio Can you run this python script and share the output?
[test_openvino.zip](https://github.com/immich-app/immich/files/14436219/test_openvino.zip)

**Reply by KoffeinKaio** [reply]
Docker immich-machine-learning logs:
```
immich_machine_learning  | [02/28/24 22:23:33] DEBUG    Checking for inactivity...                         
immich_machine_learning  | [02/28/24 22:23:39] DEBUG    Available ORT providers:                           
immich_machine_learning  |                              {'OpenVINOExecutionProvider',                      
immich_machine_learning  |                              'CPUExecutionProvider'}                            
immich_machine_learning  | [02/28/24 22:23:39] DEBUG    Available OpenVINO devices: ['CPU', 'GPU']         
immich_machine_learning  | [02/28/24 22:23:39] INFO     Setting 'ViT-B-32__openai' execution providers to  
immich_machine_learning  |                              ['OpenVINOExecutionProvider',                      
immich_machine_learning  |                              'CPUExecutionProvider'], in descending order of    
immich_machine_learning  |                              preference                                         
immich_machine_learning  | [02/28/24 22:23:39] DEBUG    Setting execution provider options to              
immich_machine_learning  |                              [{'device_type': 'GPU_FP32'},                      
immich_machine_learning  |                              {'arena_extend_strategy': 'kSameAsRequested'}]     
immich_machine_learning  | [02/28/24 22:23:39] DEBUG    Setting execution_mode to ORT_SEQUENTIAL           
immich_machine_learning  | [02/28/24 22:23:39] DEBUG    Setting inter_op_num_threads to 0                  
immich_machine_learning  | [02/28/24 22:23:39] DEBUG    Setting intra_op_num_threads to 0                  
immich_machine_learning  | [02/28/24 22:23:39] DEBUG    Setting preferred runtime to onnx                  
immich_machine_learning  | [02/28/24 22:23:39] INFO     Downloading clip model 'ViT-B-32__openai'. This may
immich_machine_learning  |                              take a while.                                      
immich_machine_learning  | [02/28/24 22:23:43] DEBUG    Checking for inactivity...                         
config.json: 100%|██████████| 196/196 [00:00<00:00, 1.20MB/s]
.gitattributes: 100%|██████████| 1.52k/1.52k [00:00<00:00, 13.0MB/s]
textual/merges.txt: 100%|██████████| 525k/525k [00:00<00:00, 2.44MB/s]
visual/preprocess_cfg.json: 100%|██████████| 197/197 [00:00<00:00, 1.78MB/s]
textual/tokenizer_config.json: 100%|██████████| 704/704 [00:00<00:00, 6.39MB/s]
textual/special_tokens_map.json: 100%|██████████| 472/472 [00:00<00:00, 3.50MB/s]
README.md: 100%|██████████| 422/422 [00:00<00:00, 3.14MB/s]0:00<?, ?B/s]
textual/vocab.json: 100%|██████████| 862k/862k [00:00<00:00, 1.53MB/s]
textual/tokenizer.json: 100%|██████████| 2.22M/2.22M [00:01<00:00, 2.13MB/s]
immich_machine_learning  | | 21.0M/254M [00:01<00:16, 13.8MB/s]54MB/s]
immich_machine_learning  | | 83.9M/254M [00:08<00:20, 8.34MB/s]
model.onnx: 100%|██████████| 254M/254M [00:28<00:00, 8.91MB/s]ng for inactivity...                         
Fetching 11 files:  45%|████▌     | 5/11 [00:32<00:43,  7.30s/it]for inactivity...                         
model.onnx: 100%|██████████| 351M/351M [00:29<00:00, 12.0MB/s] 
Fetching 11 files: 100%|██████████| 11/11 [00:33<00:00,  3.07s/it]
immich_machine_learning  | [02/28/24 22:24:14] INFO     Loading clip model 'ViT-B-32__openai' to memory    
immich_machine_learning  | [02/28/24 22:24:14] DEBUG    Loading clip vision model 'ViT-B-32__openai'       
immich_machine_learning  | [02/28/24 22:24:15] DEBUG    Loaded clip vision model 'ViT-B-32__openai'        
immich_machine_learning  | [02/28/24 22:24:15] DEBUG    Loading tokenizer for CLIP model 'ViT-B-32__openai'
immich_machine_learning  | [02/28/24 22:24:15] DEBUG    Loading model config for CLIP model                
immich_machine_learning  |                              'ViT-B-32__openai'                                 
immich_machine_learning  | [02/28/24 22:24:15] DEBUG    Loaded model config for CLIP model                 
immich_machine_learning  |                              'ViT-B-32__openai'                                 
immich_machine_learning  | [02/28/24 22:24:15] DEBUG    Loading tokenizer config for CLIP model            
immich_machine_learning  |                              'ViT-B-32__openai'                                 
immich_machine_learning  | [02/28/24 22:24:15] DEBUG    Loaded tokenizer config for CLIP model             
immich_machine_learning  |                              'ViT-B-32__openai'                                 
immich_machine_learning  | [02/28/24 22:24:15] DEBUG    Loaded tokenizer for CLIP model 'ViT-B-32__openai' 
immich_machine_learning  | [02/28/24 22:24:15] DEBUG    Loading visual preprocessing config for CLIP model 
immich_machine_learning  |                              'ViT-B-32__openai'                                 
immich_machine_learning  | [02/28/24 22:24:15] DEBUG    Loaded visual preprocessing config for CLIP model  
immich_machine_learning  |                              'ViT-B-32__openai'                   
```


Script output:
```
root@f6555aa606f5:/usr/src/app# python test.py 
Input
[name: "image"
type {
  tensor_type {
    elem_type: 1
    shape {
      dim {
        dim_param: "batch_size"
      }
      dim {
        dim_value: 3
      }
      dim {
        dim_value: 224
      }
      dim {
        dim_value: 224
      }
    }
  }
}
]
Output
[name: "image_embedding"
type {
  tensor_type {
    elem_type: 1
    shape {
      dim {
        dim_param: "Divimage_embedding_dim_0"
      }
      dim {
        dim_value: 512
      }
    }
  }
}
]
GPU Devices are ['GPU']
Using gpu device GPU
(1, 512)

```


ran in the immich-machine-learning container ghcr.io/immich-app/immich-machine-learning:main-openvino

**Reply by agrawalsourav98** [reply]
Okay. This only shows a single GPU. Is it the external GPU or integrated GPU? Also, Can you verify that the external gpu is visible inside docker?

**Reply by KoffeinKaio** [reply]
I dont have an integrated GPU, the ARC is the only one available

also when I run your script i can see the process and load via intel_gpu_top

**Reply by agrawalsourav98** [reply]
Nice. Then the code works for both types of intel GPUs. You got the smart search working right?

**Reply by KoffeinKaio** [reply]
Im not really sure. I can search for stuff, the ML container loads stuff but the images it displays are not really what I searched for

as I understood smart search i can search for "cars" and it should display photos with them right?

**Reply by agrawalsourav98** [reply]
Yeah, I think it matches for a lot of things currently and that can be a bit off in certain cases. Like it works like this, you run the model and then match the outputs with a bunch of things, so if the text I am matching with has ["a cat", "a dog", "a bus"], the model can say that 90% cat, 50% dog and 40% bus for an image of a cat. I don't know what filter values are being used but this is how the model works.
